### PR TITLE
Loweana

### PR DESCRIPTION
**Pull request by mantheys**
This is a generalized module for low-energy analysis and reconstruction. It is based on the DAQSimAna clustering algorithm and features full TPC backtracking and OpFlash collection. The output structure is similar to the one from the SolarNuAna_module but this conserves all hit information for reclustering purposes and advanced analysis. Contributions are more than welcome.

### CODE REVIEW

File: duneana/LowEAna/fcl/LowEAna.fcl [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  ClusterPreselectionNHit: 2         # Number of hits to match clusters.
+  
+  # Parameters for the adjacent cluster matching
+  AdjClusterTime:      1200          # Time window to search for adj. clusters in [tick] units.
+  AdjClusterRad:       100           # Radius to search for adj. clusters in [cm] units.
+  
+  # Parameters for the adjacent OpFlash matching
+  AdjOpFlashTime:      4500          # Time window to look for adjacent flashes in [tick] units should be one full drift length.
+  AdjOpFlashRad:       200           # Radius to search for adj. OpFlashes reconstructed in [cm] units.
+  AdjOpFlashMaxPECut:  1.01          # Cut on the maximum PE OpHit contribution to the total OpFlash PE (lower limit).
```
Comment (lpaulucc) [inline duneana/LowEAna/fcl/LowEAna.fcl:44]: This looks strange to me. The maximum PE contribution is lower than the minimum PE cut? And why is there a maximum cut?

File: duneana/LowEAna/LowEAna_module.cc [hunk 2]
```diff
@@ -0,0 +1,1367 @@
...
+  for ( size_t i = 0; i < fLabels.size(); i++){
+    Parts.push_back(ThisGeneratorParts); // For each label insert empty list
+    
+    art::Handle<std::vector<simb::MCTruth>> ThisHandle;
+    evt.getByLabel(fLabels[i], ThisHandle);
+    
+    if(ThisHandle){
+      lparticlestr = PrintInColor(lparticlestr,"\n"+fLabels[i]+" *is generated!",GetColor("green"));
+      
+      auto ThisValidHanlde = evt.getValidHandle<std::vector<simb::MCTruth>>(fLabels[i]);
```
Comment (lpaulucc) [inline duneana/LowEAna/LowEAna_module.cc:388]: There is a typo here (Hanlde). Is it working properly?

File: duneana/LowEAna/LowEAna_module.cc [hunk 3]
```diff
@@ -0,0 +1,1367 @@
...
+    art::Handle<std::vector<simb::MCTruth>> ThisHandle;
+    evt.getByLabel(fLabels[i], ThisHandle);
+    
+    if(ThisHandle){
+      lparticlestr = PrintInColor(lparticlestr,"\n"+fLabels[i]+" *is generated!",GetColor("green"));
+      
+      auto ThisValidHanlde = evt.getValidHandle<std::vector<simb::MCTruth>>(fLabels[i]); 
+      art::FindManyP<simb::MCParticle> Assn(ThisValidHanlde,evt,fGEANTLabel);            
+      
+      FillMyMaps( Parts[i], Assn, ThisValidHanlde);
```
Comment (lpaulucc) [inline duneana/LowEAna/LowEAna_module.cc:391]: Ah, I see the typo also here!

**Review by lpaulucc — COMMENTED** [review]
I wonder if the module can be further simplified. For example, I notice you calculate the flash purity but there exists a flash purity calculator in the photon backtracker algorithm. Could this be used instead? Could the PBT function be improved so that you could use it instead? Many of this kind of calculations are done in the bulk of the code, it could be beneficial to create separate functions for them (if there are no similar you can use from another piece of code)

File: duneana/LowEAna/fcl/LowEAna.fcl (hunk 1 above)
Comment (mantheys) [reply duneana/LowEAna/fcl/LowEAna.fcl:44]: Yes, you are right, the naming might not be super clear. There is a MinPE cut, and that one works as expected. The MaxPE cut is a cut on the ratio between the maximum number of PEs from an OpHit inside an OpFlash vs the total number of PEs of the flash. It's a way to cut on how "point-like" the flash is. But we should find a better name for it.

File: duneana/LowEAna/fcl/lowe_ana_dune10kt_1x2x6_legacy.fcl [hunk 4]
```diff
@@ -0,0 +1,40 @@
+#include "LowEAna.fcl"
```
[the remark this answers is not in the document]
Comment (mantheys) [reply duneana/LowEAna/fcl/lowe_ana_dune10kt_1x2x6_legacy.fcl:1]: Sure, If that is more clear we can do that.

**Comment by mantheys** [comment]
I thought I was using this function for flash purity double ThisOpFlashPur = pbt->OpHitCollectionPurity(SetGenTrackIDs, MOpHits) )

**Review by lpaulucc — APPROVED** [review]
Thank you for your answers, Sergio! Let's have this in now as it is and you can work on further improvements later.

**Pull request merged**